### PR TITLE
refactor(cli): remove `--local-server` flag and untangle root cmd logic

### PR DESCRIPTION
**Pull request by hiddeco**
- Reduce the typical parameters required to construct a (sub-)command to `config.CLIConfig` and `genericiooptions.IOStreams`.
- Properly implement `genericiooptions.IOStreams` in all places, by removing calls to e.g. `fmt.Println` while also wiring it into the `cobra.Command` using the available setters.
- Better utilization of `genericclioptions.PrintFlags` and the available default "name" printer when no output is configured and the command is printing information about operations performed on/for Kargo objects. This effectively changes output like:

  ```console
  $ kargo apply -f kargo          
  Stage Created: "kargo-advanced/ab-test-b"
  Stage Created: "kargo-advanced/prod-east"
  Project Updated: "/kargo-advanced"
  AnalysisTemplate Updated: "kargo-advanced/test"
  Stage Updated: "kargo-advanced/dev"
  Stage Updated: "kargo-advanced/staging"
  Stage Updated: "kargo-advanced/ab-test-a"
  Stage Updated: "kargo-advanced/prod"
  Stage Updated: "kargo-advanced/prod-west"
  Stage Updated: "kargo-advanced/prod-central"
  Warehouse Updated: "kargo-advanced/guestbook"
  ```

  to

  ```console
  $ kargo apply -f kargo                                   
  stage.kargo.akuity.io/ab-test-b created
  stage.kargo.akuity.io/prod-east created
  project.kargo.akuity.io/kargo-advanced updated
  analysistemplate.argoproj.io/test updated
  stage.kargo.akuity.io/dev updated
  stage.kargo.akuity.io/staging updated
  stage.kargo.akuity.io/ab-test-a updated
  stage.kargo.akuity.io/prod updated
  stage.kargo.akuity.io/prod-west updated
  stage.kargo.akuity.io/prod-central updated
  warehouse.kargo.akuity.io/guestbook updated
  ```
- Based on the discussion in #1569: removal of `--local-server` flag (which based on and the further discovery of https://github.com/akuity/kargo/issues/1569#issuecomment-1980596897, appears to have been broken for quite some time).

### CODE REVIEW

File: internal/cli/cmd/create/create.go
Comment (krancour) [inline internal/cli/cmd/create/create.go:67]: Just an observation: This seems to be universal for all commands.

File: internal/cli/util/util.go
Comment (krancour) [inline internal/cli/util/util.go:1]: Nit: "util" packages are something I try to avoid because they have a tendency to become "junk drawers" for every little thing that we can't decide where else to put it.

Note: We already have a few such packages that are "util" packages in all but name and long-term, I'd like to untangle those.

I'd rather see this in a package like `cli/io` or something, because that name will make it less of a magnet for homeless code.

Better yet...

I recognize this `cli/util` package only even exists because of the need to share certain functionality across all commands, which are _each_ in their own package...

I would propose that the packages under `internal/cli` are mostly vanity packages (_and_ micro-packages) and we would do well to collapse them into a single package, in which case, shared functionality like `SetIOStreams()` no longer requires its own package to be shared across all the commands.

I'll still say this is a "nit" because I wouldn't let it block this PR, and I think any package re-structuring could happen in a follow-up PR instead of piggy-backing on this one.

File: cmd/cli/root.go
Comment (krancour) [inline cmd/cli/root.go:79]: Not a blocker... more of a what if...

Knowing that much of this will shortly re-surface under a new `kargo server` command, what would you think about moving this code under a new, _hidden_ command and we'll work work on that as time permits and unhide it when it's ready?

Dealer's choice.

**Review by krancour — APPROVED** [review]
Had some discussion points, but nothing is blocking.

LGTM!

File: internal/cli/util/util.go
Comment (hiddeco) [reply internal/cli/util/util.go:1]: > Nit: "util" packages are something I try to avoid because they have a tendency to become "junk drawers" for every little thing that we can't decide where else to put it.

Yes, I agree and your `io` suggestion is actually a better way of dealing with this. The primary reason I put it into a separate place was because I didn't want to continue to pollute the `option` package which historically appeared to have been used as a "there is no better place for this" dump yard.

> [..] we would do well to collapse them into a single package [..]

The fear I have with this is that all sub-packages together may be a bit too much to continue to offer a clear separation of concerns, which could potentially open us up a temptation to not factor things out at all but rather move things into private functions. Which long-term is not a great solution either.

Given this, I will rename it to `io` before merge and let the idea simmer for a bit.

File: cmd/cli/root.go
Comment (hiddeco) [reply cmd/cli/root.go:79]: This is actually a great suggestion to keep the code around, even if it would abort as soon as the command is run.

I will implement this with a TODO note with a rough outline of what the future should bring.

File: internal/cli/util/util.go
Comment (krancour) [reply internal/cli/util/util.go:1]: > option package which historically appeared to have been used as a "there is no better place for this" dump yard.

Yep. One example of existing packages that are "util" in all but name.

> The fear I have with this is that all sub-packages together may be a bit too much to continue to offer a clear separation of concerns, which could potentially open us up a temptation to not factor things out at all but rather move things into private functions

I agree there is some logical limit to how much one lets a package sprawl before deciding it _should_ be broken up. That limit is for sure subjective. imho, the number of commands we're dealing with doesn't approach that limit.

As for common things being factored out into private functions (or types) -- I don't think that's _inherently_ terrible. 🤷‍♂️ I think there are a lot of concerns that are exclusive to a _class_ of problem -- e.g. the CLI options are exclusively a concern for CLI commands -- and keeping things of that nature close to the one place they're used feels reasonable. It doesn't preclude using other approaches for things deemed more broadly reusable.

> and let the idea simmer for a bit

💯

There's room for more discussion perhaps, and at any rate, I see no reason to rush into that sort of decision.

**Pull request merged**
